File: packages/agent/agent/lib/pdf.ts
import { readFile } from "node:fs/promises";

/**
 * Extração da camada de texto de um PDF.
 *
 * Determinística e sem modelo: o que sai daqui é o que está no arquivo. A
 * interpretação de layout — o que é data, o que é valor, o que é parcela —
 * fica com o extrator.
 */

export class PdfPasswordRequiredError extends Error {
  constructor() {
    super("PDF protegido por senha.");
    this.name = "PdfPasswordRequiredError";
  }
}

export type PdfPage = { page: number; text: string };

export type ExtractedPdf = {
  totalPages: number;
  pages: PdfPage[];
};

export type ExtractOptions = {
  password?: string;
  fromPage?: number;
  toPage?: number;
};

type MathWithSumPrecise = Math & {
  sumPrecise?: (values: Iterable<number>) => number;
};

/**
 * O pdf.js empacotado pelo unpdf 1.8 usa `Math.sumPrecise`, mas o Node 24
 * ainda não expõe essa API. Instalamos o fallback antes do import dinâmico do
 * unpdf para manter o contrato de engine declarado pela própria dependência.
 *
 * O algoritmo de Neumaier preserva melhor as somas de larguras e offsets do
 * PDF do que um `reduce` simples. Quando o runtime ganhar a API nativa, este
 * caminho deixa de fazer qualquer alteração global.
 */
function ensureMathSumPrecise(): void {
  const math = Math as MathWithSumPrecise;
  if (typeof math.sumPrecise === "function") return;

  Object.defineProperty(math, "sumPrecise", {
    configurable: true,
    writable: true,
    value(values: Iterable<number>): number {
      let sum = 0;
      let correction = 0;

      for (const value of values) {
        const next = sum + value;
        if (Math.abs(sum) >= Math.abs(value)) {
          correction += sum - next + value;
        } else {
          correction += value - next + sum;
        }
        sum = next;
      }

      return sum + correction;
    },
  });
}

/**
 * Carrega o PDF a partir da chave de armazenamento.
 *
 * Em desenvolvimento a chave é um caminho de arquivo; em produção é um blob do
 * Vercel, gravado com acesso PRIVADO. O ponto de troca é só este.
 *
 * Privado muda como se lê: um `fetch` cru no URL não basta mais, porque o blob
 * exige autenticação — e é exatamente esse o ponto. O SDK anexa o token do
 * store. Se o `fetch` cru voltar a funcionar num documento destes, é sinal de
 * que alguém regrediu o armazenamento para público.
 */
async function loadPdf(blobKey: string): Promise<Uint8Array> {
  if (blobKey.startsWith("http://") || blobKey.startsWith("https://")) {
    const token = process.env.BLOB_READ_WRITE_TOKEN;
    if (typeof token !== "string" || token.length === 0) {
      throw new Error(
        "BLOB_READ_WRITE_TOKEN não está definida nesta instância do agente. " +
          "Os documentos são privados e não podem ser lidos sem o token do store.",
      );
    }

    const { get } = await import("@vercel/blob");
    const result = await get(blobKey, { access: "private", token });

    if (result === null || result.statusCode !== 200 || result.stream === null) {
      throw new Error(`não foi possível baixar o documento (${result?.statusCode ?? "sem resposta"})`);
    }

    return new Uint8Array(await new Response(result.stream).arrayBuffer());
  }

  return new Uint8Array(await readFile(blobKey));
}

export async function extractPdfText(
  blobKey: string,
  options: ExtractOptions = {},
): Promise<ExtractedPdf> {
  const data = await loadPdf(blobKey);

  ensureMathSumPrecise();
  const { extractText, getDocumentProxy } = await import("unpdf");

  let pdf;
  try {
    pdf = await getDocumentProxy(data, { password: options.password });
  } catch (error) {
    if (isPasswordError(error)) throw new PdfPasswordRequiredError();
    throw error;
  }

  const { totalPages, text } = await extractText(pdf, { mergePages: false });
  const allPages: string[] = Array.isArray(text) ? text : [text];

  const from = Math.max(1, options.fromPage ?? 1);
  const to = Math.min(totalPages, options.toPage ?? totalPages);

  const pages: PdfPage[] = [];
  for (let page = from; page <= to; page += 1) {
    pages.push({ page, text: (allPages[page - 1] ?? "").trim() });
  }

  return { totalPages, pages };
}

/**
 * O pdf.js sinaliza senha por nome de exceção, não por tipo exportado, então a
 * detecção é por string. Frágil o suficiente para merecer o comentário: se um
 * PDF protegido passar batido, é aqui que se olha.
 */
function isPasswordError(error: unknown): boolean {
  if (!(error instanceof Error)) return false;
  return (
    error.name === "PasswordException" ||
    /password/i.test(error.message) ||
    /encrypted/i.test(error.message)
  );
}

File: services/agent/clara/tools/pdf.py
"""Extração da camada de texto de um PDF — porta de `agent/lib/pdf.ts`.

Determinística e sem modelo: o que sai daqui é o que está no arquivo. Usa
`pypdf` no lugar do `unpdf`/pdf.js — mesma restrição da v1: só PDFs nativos
digitais, sem OCR.
"""

from __future__ import annotations

from dataclasses import dataclass

import httpx
from pypdf import PdfReader
from pypdf.errors import FileNotDecryptedError, PdfReadError

from clara.settings import get_settings


class PdfPasswordRequiredError(Exception):
    pass


@dataclass
class PdfPage:
    page: int
    text: str


@dataclass
class ExtractedPdf:
    total_pages: int
    pages: list[PdfPage]


def _load_pdf_bytes(blob_key: str) -> bytes:
    """Em desenvolvimento a chave é um caminho de arquivo; em produção é uma
    URL do Vercel Blob, gravada com acesso PRIVADO."""
    if blob_key.startswith("http://") or blob_key.startswith("https://"):
        token = get_settings().blob_read_write_token
        if not token:
            raise RuntimeError(
                "BLOB_READ_WRITE_TOKEN não está definida nesta instância do agente. "
                "Os documentos são privados e não podem ser lidos sem o token do store."
            )
        response = httpx.get(blob_key, headers={"Authorization": f"Bearer {token}"}, timeout=30)
        response.raise_for_status()
        return response.content

    with open(blob_key, "rb") as f:
        return f.read()


def extract_pdf_text(
    blob_key: str,
    *,
    password: str | None = None,
    from_page: int | None = None,
    to_page: int | None = None,
) -> ExtractedPdf:
    data = _load_pdf_bytes(blob_key)

    try:
        reader = PdfReader(data if hasattr(data, "read") else __import__("io").BytesIO(data))
        if reader.is_encrypted:
            if password is None:
                raise PdfPasswordRequiredError()
            if not reader.decrypt(password):
                raise PdfPasswordRequiredError()
    except FileNotDecryptedError as exc:
        raise PdfPasswordRequiredError() from exc
    except PdfReadError as exc:
        if "password" in str(exc).lower() or "encrypt" in str(exc).lower():
            raise PdfPasswordRequiredError() from exc
        raise

    total_pages = len(reader.pages)
    start = max(1, from_page or 1)
    end = min(total_pages, to_page or total_pages)

    pages = [
        PdfPage(page=i, text=(reader.pages[i - 1].extract_text() or "").strip())
        for i in range(start, end + 1)
    ]
    return ExtractedPdf(total_pages=total_pages, pages=pages)
